### ISSUE TITLE
Bug in recent "OptimizeInstructions: Optimize StructNew/ArrayNew forms"

### ISSUE THREAD

**Issue by mkustermann** [WebAssembly/binaryen#6593]
It seems the recent optimization in fb5608a924dc3d53c4fb5cbd62fa5c5e6c87c707 may contain a bug where it incorrectly considers two inputs to be equal in the `visitArrayNewFixed` implementation.

Reproduction 
[m3_build_web.tar.gz](https://github.com/WebAssembly/binaryen/files/15318935/m3_build_web.tar.gz)

Extract the tarball and serve the `build/web` folder with `'--headers=Cross-Origin-Embedder-Policy=credentialless;Cross-Origin-Opener-Policy=same-origin' http headers.

=> Observe the text is rendered correctly.

Now optimize the `build/web/main.dart.wasm` file with binaryen via
```
% wasm-opt --all-features --closed-world --traps-never-happen --type-unfinalizing -Os --type-ssa --gufa -Os --type-merging -Os --type-finalizing -g build/web/main.dart.wasm -o build/web/main.dart.wasm
```
and cold-reload the page.

=> Observe the text is now rendered incorrectly

(By briefly looking at it it seems the issue is in the `$initializeEngineServices` function where an array of two different `_Future` objects turn into an array of two identical `_Future` objects)

/cc @kripken

**Comment by mkustermann** [comment]
The issue seems to still be there on `origin/main`

**Comment by kripken** [comment]
Thanks for the comprehensive bug report @mkustermann ! Fix in #6596.
